File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
private:
    int solve(int curr, int n){
        if(curr > n) {
            return 0;
        }

        if (curr == n){
            return 1;
        }

        if (m[curr] != -1) return m[curr];

        int two = solve(curr + 2, n);
        int one = solve(curr + 1, n);

        return m[curr] = two + one;
    }
public:
    int m[46];
    int climbStairs(int n) {
        int curr = 0;
        for(int i = 0; i < 46; i++){
            m[i] = -1;
        }
        return solve(curr, n);
    }
};

/*
1 + 1 + 1 + 1 + 1
1 + 1 + 1 + 2
1 + 1 + 2 + 1
1 + 2 + 1 + 1
2 + 1 + 1 + 1
1 + 2 + 2
2 + 1 + 2
2 + 2 + 1
*/

File: 0070-climbing-stairs/0070-climbing-stairs.go
func solve(curr int, n int, m []int) int {
    if curr > n {
        return 0
    }
    if curr == n {
        return 1
    }

    if m[curr] != -1 {
        return m[curr]
    }

    var two int = solve(curr + 2, n, m)
    var one int = solve(curr + 1, n, m)
    m[curr] = two + one
    return m[curr]
}

func climbStairs(n int) int {
    var curr int = 0

    m := make([]int, 46)

    for i := range m {
        m[i] = -1
    }

    return solve(curr, n, m)
}